src/commands/externalDiff.ts

```diff
@@ -0,0 +1,87 @@
+'use strict';
+import { SourceControlResourceState, Uri, window } from 'vscode';
+import { Command, Commands } from './common';
+import { CommandContext } from '../commands';
+import { GitService } from '../gitService';
+import { Logger } from '../logger';
+import { Messages } from '../messages';
+
+enum ResourceGroupType {
+    Merge,
+    Index,
+    WorkingTree
+}
+
+interface Resource extends SourceControlResourceState {
+    readonly resourceGroupType: ResourceGroupType;
+}
+
+class ExternalDiffFile {
+    constructor(public uri: Uri, public staged: boolean) {
+    }
+}
+
+export class ExternalDiffCommand extends Command {
+
+    constructor(private git: GitService) {
+        super(Commands.ExternalDiff);
+    }
+
+    protected async preExecute(context: CommandContext): Promise<any> {
+        const repoPath = await this.git.getRepoPathFromUri(undefined);
+        if (!repoPath) return Messages.showNoRepositoryWarningMessage(`Unable to open changed files`);
+
+        let files: ExternalDiffFile[];
+
+        if (context.type === 'scm-states') {
+            files = [];
+
+            context.scmResourceStates.forEach((_: Resource) => {
+                files.push(new ExternalDiffFile(_.resourceUri, _.resourceGroupType === ResourceGroupType.Index));
+            });
+
+            return this.execute(repoPath, files);
+        }
+
+        const status = await this.git.getStatusForRepo(repoPath);
```

Comment:
Possibly the same here -- maybe move this in to `execute` for when there aren't any files specified in the args